django/db/transaction.py

```diff
@@ -292,37 +324,108 @@ def __exit__(self, exc_type, exc_value, traceback):
                 else:
                     # Roll back transaction
                     try:
-                        connection.rollback()
+                        yield connection.rollback()
                     except Error:
                         # An error during rollback means that something
                         # went wrong with the connection. Drop it.
-                        connection.close()
+                        yield connection.close()
 
         finally:
             # Outermost block exit when autocommit was enabled.
             if not connection.in_atomic_block:
                 if connection.closed_in_transaction:
                     connection.connection = None
                 else:
-                    connection.set_autocommit(True)
+                    yield connection.set_autocommit(True)
             # Outermost block exit when autocommit was disabled.
             elif not connection.savepoint_ids and not connection.commit_on_exit:
                 if connection.closed_in_transaction:
                     connection.connection = None
                 else:
                     connection.in_atomic_block = False
 
+        yield
+
+
+class Atomic(ContextDecorator, _BaseAtomic):
+    __doc__ = _BaseAtomic.__doc__
+
+    def get_connection(self, using):
+        return get_connection(using)
+
+    def __enter__(self):
+        run_sync_generator(self._enter_gen())
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        run_sync_generator(self._exit_gen(exc_type, exc_value, traceback))
+
+
+class AsyncAtomic(AsyncContextDecorator, _BaseAtomic):
```

Comment:
I’m not sure how to test this without fully duplicating the test case. any ideas?